source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst

```diff
@@ -1921,6 +2089,73 @@ shard:
 When these are returned, the mongos should *not* be marked as "Unknown", since
 it is more likely an issue with the shard.
 
+Why ignore errors based on CMAP's generation number?
+''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+Using CMAP's generation number solves the following race condition among
+application threads and the monitor during error handling:
+
+#. Two concurrent writes begin on application threads A and B.
+#. The server restarts.
+#. Thread A receives the first non-timeout network error, and the client
+   marks the server Unknown, and clears the server's pool.
+#. The client re-checks the server and marks it Primary.
+#. Thread B receives the second non-timeout network error and the client
+   marks the server Unknown again.
+
+The core issue is that the client processes errors in arbitrary order
+and may overwrite fresh information about the server's status with stale
+information. Using CMAP's generation number avoids the race condition because
+the duplicate (or stale) network error can be identified (changes in
+**bold**). This change applies to both the old protocol and new streaming
```

Comment:
will we continue to have an "old protocol" and a "streaming protocol", or will we just have a "protocol"?